Library/Homebrew/cli_parser.rb

```diff
@@ -0,0 +1,280 @@
+module HomebrewCli
+  class Parser
+    attr_accessor :ignore_invalid_options
+    attr_reader :unnamed
+
+    def initialize(&block)
+      @specs = {}
+      @long = {}
+      @short = {}
+      @order = []
+      @constraints = []
+
+      instance_eval(&block)
+    end
+
+    def option(name, desc = "", opts = {})
+      opts[:desc] ||= desc
+
+      o = Option.create(name, desc, opts)
+
+      raise StandardError, "option #{o.name} already present" if @specs.member?(o.name) || @long[o.long]
+
+      @long[o.long] = o.name
+      @specs[o.name] = o
+
+      @order << [:opt, o.name]
+    end
+
+    def depends(*options)
+      options.each { |option| raise StandardError, "unknow #{option}" unless @specs[option] }
+      @constraints << [:depends, options]
+    end
+
+    def conflicts(*options)
+      options.each { |option| raise StandardError, "unknow #{option}" unless @specs[option] }
+      @constraints << [:conflicts, options]
+    end
+
+    def parse(cmdline = ARGV)
```

Comment:
I'd consider not having a default argument here to decouple things a bit. This function is also pretty long and might be a good candidate to split into a few smaller functions.

Reply:
Now this is split this into smaller functions.
Can you explain how not having a default argument would decouple things? The default argument is to prevent passing ARGV explicitly everywhere.